pkg/cloudmap/cache.go

```diff
@@ -0,0 +1,148 @@
+package cloudmap
+
+import (
+	"fmt"
+	"github.com/aws/aws-cloud-map-mcs-controller-for-k8s/pkg/model"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/util/cache"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"time"
+)
+
+const (
+	nsKeyPrefix    = "ns"
+	svcKeyPrefix   = "svc"
+	endptKeyPrefix = "endpt"
+
+	defaultCacheSize = 1024
+	defaultNsTTL     = 2 * time.Minute
+	defaultSvcTTL    = 2 * time.Minute
+	defaultEndptTTL  = 5 * time.Second
+)
+
+type ServiceDiscoveryClientCache interface {
+	GetNamespace(namespaceName string) (namespace *model.Namespace, found bool)
+	CacheNamespace(namespace *model.Namespace)
+	CacheNilNamespace(namespaceName string)
+	GetServiceId(namespaceName string, serviceName string) (serviceId string, found bool)
+	CacheServiceId(namespaceName string, serviceName string, serviceId string)
+	GetEndpoints(serviceId string) (endpoints []*model.Endpoint, found bool)
+	CacheEndpoints(serviceId string, endpoints []*model.Endpoint)
+	EvictEndpoints(serviceId string)
+}
+
+type sdCache struct {
+	log    logr.Logger
+	cache  *cache.LRUExpireCache
+	config sdCacheConfig
+}
+
+type sdCacheConfig struct {
+	nsTTL    time.Duration
+	svcTTL   time.Duration
+	endptTTL time.Duration
+}
+
+func NewDefaultServiceDiscoveryClientCache() ServiceDiscoveryClientCache {
+	return &sdCache{
+		log:   ctrl.Log.WithName("cloudmap"),
+		cache: cache.NewLRUExpireCache(defaultCacheSize),
+		config: sdCacheConfig{
+			nsTTL:    defaultNsTTL,
+			svcTTL:   defaultSvcTTL,
+			endptTTL: defaultEndptTTL,
+		}}
+}
+
+func (sdCache *sdCache) GetNamespace(nsName string) (ns *model.Namespace, found bool) {
+	key := sdCache.buildNsKey(nsName)
+	entry, exists := sdCache.cache.Get(key)
+	if !exists {
+		return nil, false
+	}
+
+	if entry == nil {
+		return nil, true
+	}
+
+	nsEntry, ok := entry.(model.Namespace)
+	if !ok {
+		sdCache.log.Info("failed to retrieve namespace from cache", "nsName", nsName)
```

Comment:
This is unexpected state, as we should only `put` namespace for this key. We should log this as an error.